src/lib.rs

```diff
@@ -40,24 +55,47 @@ impl TsnSocket {
     }
 }
 
-const CONTROL_SOCK_PATH: &str = "/var/run/tsn.sock";
-
-fn send_cmd(command: String) -> Result<String, std::io::Error> {
-    let mut stream = UnixStream::connect(CONTROL_SOCK_PATH)?;
-    stream.write_all(command.as_bytes())?;
-    let mut msg = String::new();
-    stream.read_to_string(&mut msg)?;
-    Ok(msg)
-}
-
-fn create_vlan(ifname: &str, vlanid: u16) -> Result<String, std::io::Error> {
-    let command = format!("create {} {}\n", ifname, vlanid);
-    send_cmd(command)
+fn create_vlan(ifname: &str, vlanid: u16) -> Result<i32, i32> {
+    let config = get_config(ifname);
+    let shm_name = get_shmem_name(ifname, vlanid);
+    let shm_fd = get_shmem_fd(&shm_name);
+    lock_shmem(&shm_fd);
+    let mut vlan_vec = read_shmem(&shm_name);
+    let mut result = Ok(0);
+    if vlan_vec.is_empty() {
+        result = vlan::create_vlan(&config, ifname, vlanid);
+    }
+    vlan_vec.push(process::id());
+    write_shmem(&shm_name, &vlan_vec);
+    unlock_shmem(&shm_fd);
+    result
 }
 
-fn delete_vlan(ifname: &str, vlanid: u16) -> Result<String, std::io::Error> {
-    let command = format!("delete {} {}\n", ifname, vlanid);
-    send_cmd(command)
+fn delete_vlan(ifname: &str, vlanid: u16) -> Result<i32, i32> {
+    let shm_name = get_shmem_name(ifname, vlanid);
+    let shm_fd = get_shmem_fd(&shm_name);
+    let mut result = Ok(0);
+    lock_shmem(&shm_fd);
+    let mut vlan_vec = read_shmem(&shm_name);
+
+    for i in 0..vlan_vec.len() {
+        if vlan_vec[i] == process::id() {
+            vlan_vec.remove(i);
+            break;
+        }
```

Comment:
Also checking dead pids is required here

Reply:
already check at read_shmem and locking shmem

Reply:
Then, the bug happens.

If we kill last process and launch new instance again, This happens
- As we killed last process, vlan is not destroyed
- On read_shmem, cleanup happened (pid is deleted and the vec is empty now)
- As new instance thinks they are the first instance, So try to create vlan, But it is already there

So, cleanup process is should only happens on destroy